### PR TITLE
Add live-updatable VM tolerations

### PR DESCRIPTION
**Pull request by acardace**
### What this PR does
This PR will allow updating of the VMs' node tolerations which will live update the relevant VMI fields.

To have the tolerations actually reflected on the virt-launcher pod the VM owner will have to migrate the VM, only then the tolerations will be applied. Prior to this PR a VM owner had to restart the VM to apply tolerations.

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Allow live updating VMs' tolerations 
```

### CODE REVIEW

File: tests/hotplug/tolerations.go
Comment (fossedihelm) [inline tests/hotplug/tolerations.go:124]: Changing this to:
```
func validateTolerations(virtClient kubecli.KubevirtClient, vmi *v1.VirtualMachineInstance, matcher gomegatypes.GomegaMatcher) {
	EventuallyWithOffset(1, func() []k8sv1.Toleration {
		vmi, err := virtClient.VirtualMachineInstance(vmi.Namespace).Get(context.Background(), vmi.Name, k8smetav1.GetOptions{})
		Expect(err).NotTo(HaveOccurred())
		return vmi.Spec.Tolerations
	}, 60*time.Second, time.Second).Should(matcher)
}
```
will allow you to use it with more flexibility.
`validateTolerations(virtClient, vmi, ContainElement(tolerations[0]))`
and the last Eventually loop (where you check the toleration deletion):
`validateTolerations(virtClient, vmi, Not(ContainElement(tolerations[0])))`

The question is: do we really need sematic.DeepEqual?

**Review by fossedihelm — COMMENTED** [review]
Thank you @acardace 
A couple of comments below.

File: tests/hotplug/tolerations.go
Comment (xpivarc) [inline tests/hotplug/tolerations.go:45]: `libvmi.WithNamespace(testsuite.GetTestNamespace(nil)` is smell.

File: tests/hotplug/tolerations.go
Comment (xpivarc) [inline tests/hotplug/tolerations.go:79]: This seems redundant.

File: tests/hotplug/tolerations.go
Comment (xpivarc) [inline tests/hotplug/tolerations.go:97]: This seems redundant.

File: tests/hotplug/tolerations.go
Comment (xpivarc) [reply tests/hotplug/tolerations.go:105]: Return the error.

File: tests/hotplug/tolerations.go
Comment (xpivarc) [reply tests/hotplug/tolerations.go:124]: Or just have a function that returns tolerations and you assert on caller.

File: pkg/virt-controller/watch/vm/vm_test.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm/vm_test.go:6262]: We should have a test case for:
1. add tolerations to none
2. add tolerations to some
3. remove some from some
4. remove all from some

File: tests/hotplug/tolerations.go
Comment (acardace) [reply tests/hotplug/tolerations.go:124]: No we don't you're right, I thought there was another struct inside :+1: .

File: tests/hotplug/tolerations.go
Comment (acardace) [reply tests/hotplug/tolerations.go:79]: You're right, I'll remove it.

File: tests/hotplug/tolerations.go
Comment (acardace) [reply tests/hotplug/tolerations.go:97]: why? I have to test that the toleration gets removed, or are you referring to something else?

**Comment by acardace** [comment]
@xpivarc @fossedihelm I think I addressed all comments, please take a look.

File: tests/hotplug/tolerations.go
Comment (fossedihelm) [reply tests/hotplug/tolerations.go:64]: I think that the suggestion is to move the `testsuite.GetTestNamespace(nil)` in the creation param:
```
vm, err := virtClient.VirtualMachine(testsuite.GetTestNamespace(nil)).Create(context.Background(), vm, metav1.CreateOptions{})
```
in this case you should be able to drop the `libvmi.WithNamespace` option

**Review by fossedihelm — APPROVED** [review]
Thank you!

File: tests/hotplug/tolerations.go
Comment (fossedihelm) [inline tests/hotplug/tolerations.go:71]: The test fails because at this stage the `vmi` variable has empty namespace.
You need to retrieve the correct vmi (with the right namespace) before performing this statement (OR explicitly set the namespace :D which is what  we were suggesting not to do)

Comment (acardace) [reply tests/hotplug/tolerations.go:71]: Yup, I've changed it back.

**Comment by fossedihelm** [comment]
/lgtm
Thanks

File: tests/hotplug/tolerations.go
Comment (xpivarc) [reply tests/hotplug/tolerations.go:64]: Thank you Federico :)

File: pkg/virt-controller/watch/vm/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm/vm.go:3323]: You want this in the previous commit

Comment (acardace) [reply pkg/virt-controller/watch/vm/vm.go:3323]: on it, rebasing mistake.

File: pkg/virt-controller/watch/vm/vm.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm/vm.go:790]: Should not start with capital letter.

File: pkg/virt-controller/watch/vm/vm_test.go
Comment (xpivarc) [inline pkg/virt-controller/watch/vm/vm_test.go:6280]: This is wrong as the VM should be created with `updatedTolerations`

File: tests/hotplug/tolerations.go
Comment (fossedihelm) [inline tests/hotplug/tolerations.go:69]: This is needed to avoid an additional `Get` request.
I agree that we should discourage the usage of ` libvmi.WithNamespace` when non-necessary, but
maybe in this case it was better. However, I will not block this.
/lgtm

Comment (acardace) [reply tests/hotplug/tolerations.go:69]: I agree (this was my first version), but I'll just stick to whatever is the reviewers/approvers consensus.

**Review by xpivarc — COMMENTED** [review]
/approve

File: pkg/virt-controller/watch/vm/vm.go
Comment (acardace) [reply pkg/virt-controller/watch/vm/vm.go:790]: you're right, it's because I copied it from the affinity code

File: pkg/virt-controller/watch/vm/vm_test.go
Comment (xpivarc) [reply pkg/virt-controller/watch/vm/vm_test.go:6280]: That is true as of today but we do not know about the future.

**Review by fossedihelm — APPROVED** [review]

**Review by fossedihelm — APPROVED** [review]

**Comment by fossedihelm** [comment]
/hold
rebase needed

**Pull request merged**
